ark/schema/structure/structure.ts

```diff
@@ -812,6 +831,43 @@ const getPossibleMorph = (
 			for (const k in data) if (!node.declaresKey(k)) delete (data as dict)[k]
 
 		return data
+	}
+
+	return (defaultableMorphsCache[cacheKey] = $arkStructuralMorph)
+}
+
+const precompileMorphs = (js: NodeCompiler, node: Structure.Node) => {
+	const requiresContext =
+		node.defaultable.some(node => node.defaultValueMorph.length === 2) ||
+		node.sequence?.defaultValueMorphs.some(morph => morph.length === 2)
+
+	const args = `(data${requiresContext ? ", ctx" : ""})`
+
+	return js.block(`${args} => `, js => {
+		for (let i = 0; i < node.defaultable.length; i++) {
+			const { serializedKey, defaultValueMorphRef } = node.defaultable[i]
+			js.if(`!(${serializedKey} in data)`, js =>
+				js.line(`${defaultValueMorphRef}${args}`)
+			)
+		}
+
+		if (node.sequence?.defaultables) {
+			js.for(
+				`i < ${node.sequence.defaultables.length}`,
+				js => js.set(`data[i]`, 5),
+				`data.length - ${node.sequence.prefixLength}`
+			)
+		}
```

Comment:
@ssalbdivad Hard-coded `5`? Strange this hasn't been noticed since March. 

```ts
  if (node.sequence?.defaultables) {
      js.for(
          `i < ${node.sequence.defaultables.length}`,
          js => js.line(`${node.sequence.defaultValueMorphsReference}[i]${args}`),
          `data.length - ${node.sequence.prefixLength}`
      )
  }
```